### PR TITLE
Guide: introduction to working with data in CUBA - #1

### PR DESCRIPTION
**Pull request by mariodavid**
This PR creates a guide for "Introduction to working with data in CUBA" (#1)

### CODE REVIEW

**Comment by mariodavid** [comment]
@oshiryaeva can you accept the CLA? ;) thanks

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+It also adds certain capabilities that are not available out of the box in JPA. Those capabilities are mainly the https://doc.cuba-platform.com/manual-6.10/views.html[Views concept] as well as the application of https://doc.cuba-platform.com/manual-6.10/security_subsystem.html[security constrains]. More information about differences between JPA and DataManager can be found in the docs on https://doc.cuba-platform.com/manual-6.10/dm_vs_em.html[DataManager vs. EntityManager].
+
+
+include::includes/petclinic_introduction.adoc[]
+
+
+== Load data with DataManager
+
+DataManager has multiple ways to load data from the database. The following examples contain the most common cases for loading data in different ways.
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Sounds a bit like a tautology. Maybe rephrase?

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+The part that loads the Pet instance by its ID looks like this with the use of the DataManager API:
+
+
+.VisitServiceBean.java
+[source,java]
+----
+@Inject
+DataManager dataManager;
+
+Pet loadPetById(String petId) {
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Is there any reason to make this method package-private?

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  return dataManager.load(Pet.class)
+            .query("select e from petclinic$Pet e where e.petId = :petId")
+            .parameter("petId", petId)
+            .one();
+}
+----
+
+DataManager has a https://martinfowler.com/bliki/FluentInterface.html[fluent API] which allows to chain operations on the load operation in order to define parameters to the execution. In this case, the interaction starts with `dataManager.load(Pet.class)`. Then the JPQL query is defined and the associated parameters are configured with the parameter values. The method call `one()` executes the load operation and fetches exactly one item that matches the criteria. If there are no results or multiple results the corresponding exceptions are thrown.
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Method `one()` doesn't throw exceptions in case of multiple results returned by the query. It just returns the first available.

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+It is up to the application developer to decide if the query can return one or possibly multiple results. In this case it is only possible to create a Pet ID once, we can be sure that the query will return at the most one item.
+
+In this case, the attribute `petId` is not the primary key of the entity. If this would be the case, the query would be even easier to create as the query part is not needed: `dataManager.load(Pet.class).id(petId).one();`
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Please explain here what `petId` actually is.

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+=== Multiple Items by query
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Multiple Items by Query

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+The next very common option is to load multiple Items defined by a query.
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: ... multiple items...

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                .parameter("ownerCity", city)
+                .parameter("petType", petType)
+                .view("pet-with-owner-and-type")
+                .list();
+}
+----
+
+The query part is a little more complicated compared to above, but the same pattern applies. The differences in this example are the usage of the `list()` method as we expect multiple Pets from the database query. Also a specific view has been used, because not only the Pet data, but also the Owner data has to be loaded in order to send out the Mailings.
+
+== Write data with DataManager
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Write Data with DataManager
plz make sure all headings are in Title Case

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+----
+
+<1> an entity instance is created with all initializations in memory
+<2> the entity instance is stored in the DB
+
+The creation of an instance through DataManager does the initialization of the instance as well. It is important to not create an instance via `new Visit()` but rather through this Factory method, so that CUBA can do all the required initialization (like `@PostConstruct` annotated methods e.g.).
+
+Once the entity instance was created, the data of that entity has to be set through its corresponding setter methods. In this case, the database instance of the Pet is set as well as the current date as the visit date.
+
+The process of storing an entity instance contains a few steps like security checks, ID generation, and so on. The return value of the commit method is the stored instance with all properties set that have been added during the storage phase.
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: In most cases (excluding IDENTITY-based entities) the ID generation happens on stage 1 in `dataManager.create(Visit.class)`

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+TIP: There are a couple over other convenience API methods for storing entites. For more information about those variants see the http://files.cuba-platform.com/javadoc/cuba/{cuba_version}/com/haulmont/cuba/core/global/DataManager.html[DataManager API reference].
```
Comment (knstvk) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: "There is a couple of other..."

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 3 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: 👍

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 2 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: no - it is actually an old snippet. Will update with "IdentificationNumber" example

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 1 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: true. will change.

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 4 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: as written above - changed it to identificationNumber - will explain it in there as well.

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 8 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: removed it and replaced it with Entity Listener execution

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 7 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: I wasn't aware of that this is common in english. It this really the case?

File: _drafts/intro-to-working-with-data-in-CUBA.adoc (hunk 9 above)
Comment (mariodavid) [reply _drafts/intro-to-working-with-data-in-CUBA.adoc]: interesting. In german it would always be "there are". When I read this here - you might be right: https://english.stackexchange.com/questions/50435/there-are-a-couple-of-apples-or-there-is-a-couple-of-apples

File: _drafts/intro-to-working-with-data-in-CUBA.adoc [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+[source,java]
+----
+@Inject
+DataManager dataManager;
+
+private Visit createVisitForPet(Pet pet) {
+
+    Visit visit = dataManager.create(Visit.class); // <1>
+
+    visit.setPet(pet);
```
Comment (mariodavid) [inline _drafts/intro-to-working-with-data-in-CUBA.adoc]: Thanks @knstvk for the review. You can feel free to just change it directly in the code if you think that it is equally fast or faster then writing the comment. There is actually a new feature in github to use suggestions in code reviews that will streamline the process.

Example of a suggestion: 
```suggestion
    visit.setPet(null);
```
